File: Coding Interviews/01 - Interview Flow/E175.InvertBinaryTree.py
"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""

class Solution:
    """
    @param root: a TreeNode, the root of the binary tree
    @return: nothing
    """
    def invertBinaryTree(self, root):
        # write your code here

        # method 2:
        # if not root:
        #     return root

        # queue = [root]
        # while queue:
        #     node = queue.pop()
        #     node.left, node.right = node.right, node.left

        #     if node.left:
        #         queue.append(node.left)
        #     if node.right:
        #         queue.append(node.right)

        # return root

        # method 1:
        # if not root:
        #     return root

        # leftnode = root.left
        # root.left, root.right = root.right, leftnode

        # self.invertBinaryTree(root.left)
        # self.invertBinaryTree(root.right)

        # return root

        # method 3:
        if not root:
            return root

        root.left, root.right = self.invertBinaryTree(root.right), self.invertBinaryTree(root.left)
        return root

File: Coding Interviews/01 - Interview Flow/E175.InvertBinaryTree.java
/**
 * Definition of TreeNode:
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left, right;
 *     public TreeNode(int val) {
 *         this.val = val;
 *         this.left = this.right = null;
 *     }
 * }
 */

public class Solution {
    /**
     * @param root: a TreeNode, the root of the binary tree
     * @return: nothing
     */
    public void invertBinaryTree(TreeNode root) {
        // write your code here
        // method 1:
        // if (root == null) return;

        // TreeNode node = root.left;
        // root.left = root.right;
        // root.right = node;

        // invertBinaryTree(root.left);
        // invertBinaryTree(root.right);

        helper(root);
    }

    private TreeNode helper(TreeNode root) {
        if (root == null) return root;

        TreeNode leftNode = helper(root.left);
        TreeNode rightNode = helper(root.right);

        root.left = rightNode;
        root.right = leftNode;

        return root;
    }
}
